vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,421 @@
+    #[error("Failed to read from guest memory: {0}")]
+    GuestMemoryRead(#[from] GuestMemoryError),
+    #[error("Arithmetic overflow in {0}")]
+    Overflow(&'static str),
+    #[error("Expected used log buffer size < {0:#x}, got: {1:#x}")]
+    BadUsedBufferSize(u32, u32),
+    #[error("Expected accumulated message length < {0:#x}, got: {1:#x}")]
+    BadAccumulatedMessageLength(u16, u16),
+}
+
+/// Definition of the diagnostics services state
+#[derive(Inspect)]
+pub struct DiagnosticsServices {
+    /// The guest physical address of the diagnostics buffer
+    gpa: Option<u32>,
+    /// Flag to indicate if we have already processed the buffer
+    did_process: bool,
+}
+
+impl DiagnosticsServices {
+    /// Create a new instance of the diagnostics services
+    pub fn new() -> DiagnosticsServices {
+        DiagnosticsServices {
+            gpa: None,
+            did_process: false,
+        }
+    }
+
+    /// Reset the diagnostics services state
+    pub fn reset(&mut self) {
+        self.gpa = None;
+        self.did_process = false;
+    }
+
+    /// Set the GPA of the diagnostics buffer
+    pub fn set_gpa(&mut self, gpa: u32) {
+        self.gpa = match gpa {
+            0 => None,
+            _ => Some(gpa),
+        }
+    }
+
+    /// Process the diagnostics buffer
+    pub fn process_diagnostics<F>(
+        &mut self,
+        gm: &GuestMemory,
+        mut log_handler: F,
+    ) -> Result<(), DiagnosticsError>
+    where
+        F: FnMut(EfiDiagnosticsLog<'_>),
+    {
+        // Validate the GPA
+        let gpa = match self.gpa {
+            Some(gpa) if gpa != 0 && gpa != u32::MAX => gpa,
+            Some(invalid_gpa) => return Err(DiagnosticsError::BadGpa(invalid_gpa)),
+            None => return Err(DiagnosticsError::NoGpa),
+        };
+
+        // Read and validate the header from the guest memory
+        let header: AdvancedLoggerInfo = gm.read_plain(gpa as u64)?;
+
+        let signature = header.signature;
+        if signature != u32::from_le_bytes(SIG_HEADER) {
+            return Err(DiagnosticsError::HeaderSignatureMismatch(
+                u32::from_le_bytes(SIG_HEADER),
+                signature,
+            ));
+        }
+
+        if header.log_buffer_size > MAX_LOG_BUFFER_SIZE {
+            return Err(DiagnosticsError::HeaderBufferSize(
+                MAX_LOG_BUFFER_SIZE,
+                header.log_buffer_size,
+            ));
+        }
+
+        // Calculate the used portion of the log buffer
+        let log_current_offset = header.log_current_offset;
+        let log_buffer_offset = header.log_buffer_offset;
+
+        let used_log_buffer_size = log_current_offset
+            .checked_sub(log_buffer_offset)
+            .ok_or_else(|| DiagnosticsError::Overflow("used_log_buffer_size"))?;
+
+        // Early exit if there is no buffer to process
+        if used_log_buffer_size == 0 {
+            tracelimit::info_ratelimited!("Used log buffer size is zero, ending processing");
```

Comment:
it might be better to trace "uefi log buffer used size is zero, ..." instead since at a glance no idea what buffer we're talking about